### DISCUSSION TITLE
Adding `tid` and `record-key` string formats to Lexicon language

### DISCUSSION THREAD

**Discussion by bnewbold** [bluesky-social/atproto#2387] [Dev Announcements]
We are making a relatively small change to the Lexicon schema language, adding two new string formats:

- `tid`: generic TID identifier, as specified [here](https://atproto.com/specs/record-key#record-key-type-tid)
- `record-key`: the generic (most flexible) version of Record Key syntax, as specified [here](https://atproto.com/specs/record-key#record-key-syntax)

We intend to use these new formats in existing Lexicons in the near future. This is technically breaking the Lexicon stability/compatibility rules, and are part of some protocol tech debt / cleanups around stabilizing atproto and the existing Lexicons ([discussion and details](https://github.com/bluesky-social/atproto/discussions/2128)).

See also:

- identifier test examples (both valid and invalid): https://github.com/bluesky-social/atproto/tree/main/interop-test-files/syntax
- specification PR to add these formats: https://github.com/bluesky-social/atproto-website/pull/248
- PR to use these formats in existing Lexicons: https://github.com/bluesky-social/atproto/pull/2378
- PR adding support in the TypeScript implementation: https://github.com/bluesky-social/atproto/pull/2377

**Comment by TahomaSoft** [comment]
Hi, would this be a valid TID (16L55C:1Z5:4KL)?

Thanks,

Erik

**Reply by DavidBuchanan314** [reply]
nope (see linked spec)

**Reply by TahomaSoft** [reply]
Thank you.  Would it though be a valid record-key?

**Reply by bnewbold** [reply]
@TahomaSoft you can play with code to check whether strings are or are not valid syntax here:
https://play.golang.com/p/imLamcQJ4Id

**Reply by TahomaSoft** [reply]
Thank you!

**Comment by TahomaSoft** [comment]
I just opened up a repository for a little code to generate a TID in python.

See:
https://github.com/TahomaSoft/Zulki-Blue-Iceflake-Record-TID
Comments and contributions welcome!

**Reply by snarfed** [reply]
Nice! You may also be interested in [arroba](https://github.com/snarfed/arroba)'s TID utilities. They're Python, mature, tested, and have been used in production for a while now:
* https://arroba.readthedocs.io/en/latest/source/arroba.html#arroba.util.datetime_to_tid
* https://github.com/snarfed/arroba/blob/e3aba1fb76c074aae8cad7e1451234bb976a84b8/arroba/util.py#L108-L201

**Reply by TahomaSoft** [reply]
Cool, thank you!  I went searching for something like this online, but could not find it.

Definitely will look at in detail!

Gracias Senor!
